File: PL_Java/PL2.java
import java.util.*;

public class PL2 {
    static Deque<Character> token = new LinkedList<>();
    static boolean zerodivde = true;
    static int aexpr()
    {
        int left = term();

        while(true)
        {
            if (token.size() != 0 && (token.peek() == '*' || token.peek() == '/'))
            {
                char op = token.peek();
                token.poll();

                int right = term();

                if (op == '*')
                {
                    left *= right;
                }
                else
                {
                    if (right == 0)
                    {
                        zerodivde = false;
                    }
                    else
                    {
                        left /= right;
                    }
                }
            }
            else
            {
                break;
            }
        }

        return left;
    }

    static int term()
    {
        int left = factor();

        while(true)
        {
            if (token.size() != 0 && (token.peek() == '+' || token.peek() == '-'))
            {
                char op = token.peek();
                token.poll();

                int right = factor();

                if (op == '+')
                {
                    left += right;
                }
                else
                {
                    left -= right;
                }
            }
            else
            {
                break;
            }
        }

        return left;
    }

    static int factor()
    {
        int num = 0;

        if (token.size() != 0 && '0' <= token.peek() && token.peek() <= '9')
        {
            return number();
        }
        else
        {
            token.poll();
            num = aexpr();

            token.poll();
            return num;
        }
    }

    static int number()
    {
        int num = 0;

        while (true)
        {
            if (token.size() != 0 && '0' <= token.peek() && token.peek() <= '9')
            {
                if (num == 0)
                {
                    num += dec();
                }
                else
                {
                    num *= 10;
                    num += dec();
                }
            }
            else
            {
                break;
            }
        }

        return num;
    }

    static int dec()
    {
        int num = token.peek() - '0';
        token.poll();

        return num;
    }

    public static void main(String[] args) {

        while (true)
        {
            Scanner sc = new Scanner(System.in);
            String syntax = sc.nextLine();

            if (syntax.length() == 0)
                break;

            boolean b = true;
            Stack<Integer> stack = new Stack<>();
            token.clear();
            zerodivde = true;

            //토큰화 및 syntax error 확인
            {
                for (int i = 0; i < syntax.length(); i++)
                {
                    char ch = syntax.charAt(i);

                    if (('(' <= ch && ch <= '+') || ch == '-' || ('/' <= ch && ch <= '9'))
                    {
                        if (ch == '+' || ch == '-' || ch == '*' || ch == '/')
                        {
                            if (token.size() == 0 || token.peekLast() == '+' || token.peekLast() == '-'
                                    || token.peekLast() == '*' || token.peekLast() == '/')
                            {
                                b = false;
                                break;
                            }
                        }
                        else if (ch == '(')
                        {
                            stack.push(0);

                            if (token.size() != 0 && ('0' <= token.peekLast() && token.peekLast() <= '9'))
                            {
                                b = false;
                                break;
                            }
                        }
                        else if (ch == ')')
                        {
                            if (stack.empty())
                            {
                                b = false;
                                break;
                            }
                            else
                            {
                                stack.pop();
                            }

                            if (token.size() != 0 && token.peekLast() == '(')
                            {
                                b = false;
                                break;
                            }
                        }
                        else
                        {
                            if (token.size() != 0 && token.peekLast() == ')')
                            {
                                b = false;
                                break;
                            }
                        }

                        token.add(ch);
                    }
                    else if (ch == ' ')
                    {
                        continue;
                    }
                    else
                    {
                        b = false;
                        break;
                    }
                }

                if (stack.size() != 0)
                {
                    b = false;
                }

                if (token.size() != 0 && (token.peekLast() == '+' || token.peekLast() == '-'
                        || token.peekLast() == '*' || token.peekLast() == '/'))
                {
                    b = false;
                }
            }

            if (b)
            {
                int res = aexpr();

                if (zerodivde)
                {
                    System.out.println(res);
                }
                else
                {
                    System.out.println("can't divide 0");
                }
            }
            else
            {
                System.out.println("syntax error!");
            }
        }
    }
}

File: PL2_python/PL2.py
from queue import Queue
from typing import List
from sys import stdin

token: Queue[str] = Queue()
zerodivide: bool = True

def aexpr() -> int:
    left: int = term()

    while True:
        if not token.empty() and (token.queue[0] == '*' or token.queue[0] == '/'):
            op: str = token.get()

            right: int = term()

            if op == '*':
                left = left * right
            else:
                if right == 0:
                    global zerodivide
                    zerodivide = False
                else:
                    left = int(left / right)
        else:
            break

    return left

def term() -> int:
    left: int = factor()

    while True:
        if not token.empty() and (token.queue[0] == '+' or token.queue[0] == '-'):
            op: str = token.get()

            right: int = factor()

            if op == '+':
                left = left + right
            else:
                left = left - right
        else:
            break
    return left

def factor() -> int:
    num: int = 0

    if not token.empty() and '0' <= token.queue[0] <= '9':
        return number()
    else:
        token.get()
        num = aexpr()

        token.get()
        return num

def number() -> int:
    num: int = 0

    while True:
        if not token.empty() and '0' <= token.queue[0] <= '9':
            if num == 0:
                num = num + dec()
            else:
                num = num * 10
                num = num + dec()
        else:
            break

    return num

def dec() -> int:
    num: int = ord(token.get()) - ord('0')

    return num

while True:
    syntax: str = stdin.readline().rstrip('\n')

    if not syntax:
        break

    b: bool = True
    stack: List[int] = []
    token = Queue()
    zerodivide = True

    # Tokenization and Syntax Error checking
    for i in range(len(syntax)):
        if ('(' <= syntax[i] <= '+') or (syntax[i] == '-') or ('/' <= syntax[i] <= '9'):
            if syntax[i] in ['+', '-', '*', '/']:
                if token.empty() or (token.queue[-1] in ['+', '-', '*', '/']):
                    b = False
                    break
            elif syntax[i] == '(':
                stack.append(0)

                if (not token.empty()) and ('0' <= token.queue[-1] <= '9'):
                    b = False
                    break
            elif syntax[i] == ')':
                if not stack:
                    b = False
                    break
                else:
                    stack.pop()

                if (not token.empty()) and token.queue[-1] == '(':
                    b = False
                    break
            else:
                if (not token.empty()) and token.queue[-1] == ')':
                    b = False
                    break

            token.put(syntax[i])
        elif syntax[i] == ' ':
            continue
        else:
            b = False
            break

    if stack or (not token.empty() and (token.queue[-1] in ['+', '-', '*', '/'])):
        b = False

    if b:
        res = aexpr()

        if zerodivide:
            print(res)
        else:
            print("can't divide 0")
    else:
        print("syntax error!")